huxley/core/models.py

```diff
@@ -273,28 +275,54 @@ def add(committee, country, school, rejected):
         def remove(assignment_data):
             deletions.append(assignment_data['id'])
 
-        for committee, country, school, rejected in new_assignments:
+        for committee, country, school, rejected, is_valid in new_assignments:
+            # If one or more fields in the assignment are invalid, the
+            # the values are passed in as CharFields instead of Model ids
             key = (committee, country)
-            old_assignment = assignment_dict.get(key)
-
-            if not old_assignment:
-                add(committee, country, school, rejected)
+            if key in assigned:
+                # Make sure that the same committee/country pair is not being
+                # given to more than one school in the upload
+                committee = str(Committee.objects.get(pk=committee).name) + ' - DUPLICATE ASSIGNMENT IN CSV'
+                country = str(Country.objects.get(pk=country).name) + ' - DUPLICATE ASSIGNMENT IN CSV'
+                school = str(School.objects.get(pk=school).name)
+                failed_assignments.append(str((school, committee, country, rejected)))
                 continue
+            elif is_valid:
+                assigned.add(key)
 
-            if old_assignment['school_id'] != school:
-                # Remove the old assignment instead of just updating it
-                # so that its delegates are deleted by cascade.
-                remove(old_assignment)
-                add(committee, country, school, rejected)
+            if is_valid:
+                old_assignment = assignment_dict.get(key)
+
+                if not old_assignment:
+                    add(committee, country, school, rejected)
+                    continue
 
-            del assignment_dict[key]
+                if old_assignment['school_id'] != school:
+                    # Remove the old assignment instead of just updating it
+                    # so that its delegates are deleted by cascade.
+                    remove(old_assignment)
+                    add(committee, country, school, rejected)
 
-        for old_assignment in assignment_dict.values():
-            remove(old_assignment)
+                del assignment_dict[key]
+            else:
+                if not Committee.objects.filter(name=committee).exists():
+                    committee = committee + ' - INVALID'
```

Comment:
Since we're checking for existence here, might be better to inform the user that that committee/country/school does not exist rather than that it's invalid